cachito/web/models.py

```diff
@@ -186,6 +186,30 @@ def to_purl(self):
             # see https://github.com/package-url/purl-spec/issues/63 for further reference
             purl_name = self.name.replace("/", "%2F")
             return f"pkg:golang/{purl_name}@{self.version}"
+
+        elif self.type == "npm":
+            purl_name = self.name.replace("@", "%40")
+            match = re.match(r"(?P<protocol>[^:]+):(?://)?(?P<suffix>.+)", self.version)
+            if not match:
+                return f"pkg:npm/{purl_name}@{self.version}"
+            protocol = match.group("protocol")
+            suffix = match.group("suffix")
+            if protocol in ("github", "gitlab", "bitbucket"):  # github:namespace/name#ref
+                match_forge = re.match(
+                    r"(?P<namespace>[^/]+)/(?P<name>[^#]+)#(?P<version>.+)$", suffix
+                )
+                if not match_forge:
+                    raise ValueError(f"Could not convert version {self.version} to purl")
```

Comment:
FYI, raising this in the API will cause a 500 Internal Server Error response. I think that's okay since this is a sanity check that should never occur.

Reply:
Agreed. In this case it would be better to halt the request execution and set it to a failed state.